source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -118,15 +118,16 @@ void Filter::setEncoderFilterCallbacks(Http::StreamEncoderFilterCallbacks& callb
 }
 
 Filter::StreamOpenState Filter::openStream() {
-  ENVOY_BUG(!processing_complete_, "openStream should not have been called");
+  // If processing already completes, no need to open a gRPC stream.
+  if (processing_complete_) {
+    // Stream failed while starting and either onGrpcError or onGrpcClose was already called
+    return sent_immediate_response_ ? StreamOpenState::Error : StreamOpenState::IgnoreError;
+  }
   if (!stream_) {
     ENVOY_LOG(debug, "Opening gRPC stream to external processor");
     stream_ = client_->start(*this, grpc_service_, decoder_callbacks_->streamInfo());
     stats_.streams_started_.inc();
-    if (processing_complete_) {
-      // Stream failed while starting and either onGrpcError or onGrpcClose was already called
-      return sent_immediate_response_ ? StreamOpenState::Error : StreamOpenState::IgnoreError;
-    }
+
```

Comment:
Thanks for the comments.
It looks to me start() function call won't change processing_complete_. If that's the case, then checking processing_complete_ is irrelevant to start() call, thus has to be moved outside to the front.

Reply:
I don't think so.  
1. Independent of this specific code, I think we should not return `StreamOpenState::Ok` if stream was failed to be opened.
2. I think it is set because `onRemoteClose` will be invoked if the stream fail to start and `processing_complete_` is set there.

It is just my pure code reading. I think we need to confirm here.

Reply:
Thanks for the good comments @tyxia !


Yes, agrees both.  We need processing_complete_ check at the beginning to deal with the situation the processing already completes, no need to open another gRPC stream.

Also, if processing_complete_ is false in the beginning, onRemoteClose()->onGrpcError() can possibly cause processing_complete_ set to be true during client->start(), thus we need keep processing_complete_ check there.  This is tested by a couple of tests in ordering_test.cc. One is here:

https://github.com/envoyproxy/envoy/blob/cab680b0ef801e269f9a50cd745728d4ae8b86e6/test/extensions/filters/http/ext_proc/ordering_test.cc#L881